presto-rcfile/src/main/java/com/facebook/presto/rcfile/RcFileReader.java

```diff
@@ -0,0 +1,584 @@
+            if (index >= 0) {
+                long blockOffset = offset + position + index + SIZE_OF_LONG + SIZE_OF_LONG;
+                input.setPosition(blockOffset);
+
+                return;
+            }
+        }
+        closeQuietly();
+    }
+
+    private void closeQuietly()
+    {
+        try {
+            close();
+        }
+        catch (IOException ignored) {
+        }
+    }
+
+    private RcFileCorruptionException corrupt(String messageFormat, Object... args)
+    {
+        closeQuietly();
+        return new RcFileCorruptionException(messageFormat, args);
+    }
+
+    private static class Column
+    {
+        private final RcFileCompressionCodec codec;
+
+        private BasicSliceInput lengthsInput;
+        private Slice dataBuffer;
+        private int uncompressedDataSize;
+
+        private byte[] decompressedBuffer = new byte[0];
+
+        private boolean compressed;
+
+        private int currentPosition;
+
+        private int currentOffset;
+        private int runLength;
+        private int lastValue;
+
+        public Column(RcFileCompressionCodec codec)
+        {
+            this.codec = codec;
+        }
+
+        private void setBuffers(Slice lengthsBuffer, Slice dataBuffer, int uncompressedDataSize)
+        {
+            this.lengthsInput = lengthsBuffer.getInput();
+            this.dataBuffer = dataBuffer;
+            this.uncompressedDataSize = uncompressedDataSize;
+
+            compressed = (codec != null);
+
+            currentPosition = 0;
+            currentOffset = 0;
+            runLength = 0;
+            lastValue = 0;
+        }
+
+        public ColumnData readColumnData(int position, int size)
+                throws IOException
+        {
+            checkArgument(size > 0 && size <= ColumnData.MAX_SIZE, "Invalid size");
+            checkArgument(currentPosition <= position, "Invalid position");
+
+            if (currentPosition < position) {
+                skipTo(position);
+            }
+
+            // read offsets
+            int[] offsets = readOffsets(size);
+
+            return new ColumnData(offsets, getDataBuffer());
+        }
+
+        private int[] readOffsets(int batchSize)
+        {
+            int[] offsets = new int[batchSize + 1];
+            offsets[0] = currentOffset;
+            for (int i = 0; i < batchSize; i++) {
+                int value;
+                if (runLength > 0) {
+                    value = lastValue;
+                    runLength--;
+                }
+                else {
+                    int length = Ints.checkedCast(readVInt(lengthsInput));
+                    if (length < 0) {
+                        runLength = (~length) - 1;
+                        value = lastValue;
+                    }
+                    else {
+                        runLength = 0;
+                        value = length;
+                    }
+                }
```

Comment:
I would abstract these lines into a function. The function would take no arguments and return `value`.
```
                int value;
                if (runLength > 0) {
                    value = lastValue;
                    runLength--;
                }
                else {
                    int length = Ints.checkedCast(readVInt(lengthsInput));
                    if (length < 0) {
                        runLength = (~length) - 1;
                        value = lastValue;
                    }
                    else {
                        runLength = 0;
                        value = length;
                    }
                }
                lastValue = value;
```